src/backend/distributed/executor/local_executor.c

```diff
@@ -617,9 +667,54 @@ SplitLocalAndRemotePlacements(List *taskPlacementList, List **localTaskPlacement
  * case of DML.
  */
 static uint64
-ExecuteLocalTaskPlan(PlannedStmt *taskPlan, char *queryString,
-					 TupleDestination *tupleDest, Task *task,
-					 ParamListInfo paramListInfo)
+LocallyExecuteTaskPlan(PlannedStmt *taskPlan, char *queryString,
+					   TupleDestination *tupleDest, Task *task,
+					   ParamListInfo paramListInfo)
+{
+	volatile uint64 processedRows = 0;
+
+	/*
+	 * If we roll back to a savepoint, we may no longer be in a query on
+	 * a shard. Reset the value as we go back up the stack.
+	 */
+	bool prevInLocalQueryOnShard = InLocalQueryOnShard;
+
+	if (task->anchorShardId != INVALID_SHARD_ID)
+	{
+		InLocalQueryOnShard = true;
+	}
+
+	LocalExecutorLevel++;
```

Comment:
We could maybe do without the LocalExecutorLevel now and use InLocalQueryOnShard instead. It is mainly used to allow us to execute drop and truncate tasks on shards from the local executor, while we prevent that for regular users. Those tasks would have an anchor shard, so InLocalQueryOnShard would be true.